### ISSUE TITLE
Injection fails with StatefulSets

### ISSUE THREAD

**Issue by darkn3rd** [hashicorp/consul-k8s#1523]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request. Searching for pre-existing feature requests helps us consolidate datapoints for identical requirements into a single place, thank you!
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### Overview of the Issue

Consul fails with StatefulSets.  No documentation on this common use case.

### Reproduction Steps

1. Deploy Consul as per docs:
   ```bash
   helm install consul hashicorp/consul \
     --create-namespace \
     --namespace consul \
     --set global.name=consul \
     --set connectInject.enabled=consul \
     --set controller.enabled=consul
   ```
2. Deploy application that uses StatefulSet:
   ```bash
   cat << EOF > dgraph_config.yaml
   image:
     tag: v21.03.2
   zero:
     extraAnnotations:
       consul.hashicorp.com/connect-inject: 'true'
       consul.hashicorp.com/transparent-proxy: 'true'
   alpha:
     extraAnnotations:
       consul.hashicorp.com/connect-inject: 'true'
       consul.hashicorp.com/transparent-proxy: 'true'
     configFile:
       config.yaml: |
         security:
           whitelist: {{ env "DG_ACCEPT_LIST" | default "0.0.0.0/0" | quote }}
   EOF
   helm repo add dgraph https://charts.dgraph.io
   helm install dgraph dgraph/dgraph --values dgraph.yaml
   ```
3. Observe failure, check logs of containers.

### Logs

* `consul-connect-inject-init` container:
  ```
  2022-09-22T06:37:22.601Z [ERROR] There are multiple Consul services registered for this pod when there must only be one. Check if there are multiple Kubernetes services selecting this pod and add the label `consul.hashicorp.com/service-ignore: "true"` to all services except the one used by Consul for handling requests.
  2022-09-22T06:37:23.604Z [INFO]  Unable to find registered services; retrying
  ```
* `envoy-sidecar` container:
  ```
  Failed to load logs: previous terminated container "envoy-sidecar" in pod "dgraph-dgraph-alpha-0" not found
  Reason: BadRequest (400)
  ```
* `copy-consul-bin` container:
  ```
  Failed to load logs: previous terminated container "copy-consul-bin" in pod "dgraph-dgraph-alpha-0" not found
  Reason: BadRequest (400)
  ```

### Expected behavior

I expected the solution to work after following documentation from: https://www.consul.io/docs/k8s/installation/install

### Environment details

- `consul-k8s` version: v0.48.0
- `values.yaml` used to deploy the helm chart:
  ```yaml
  global:
    name: consul
  connectInject:
    enabled: true
  controller:
    enabled: true
  ```
- Kubernetes version: v1.22.12-gke.300
- Cloud Provider: GKE
- Networking CNI plugin in use: whatever comes default with GKE

### Additional Context

These are the services created for Dgraph.  The headless services are required for Dgraph for cluster members to communicate to each other:

```
dgraph-dgraph-alpha            ClusterIP   10.108.9.194   <none>        8080/TCP,9080/TCP   39m
dgraph-dgraph-alpha-headless   ClusterIP   None           <none>        7080/TCP            39m
dgraph-dgraph-zero             ClusterIP   10.108.2.33    <none>        5080/TCP,6080/TCP   39m
dgraph-dgraph-zero-headless    ClusterIP   None           <none>        5080/TCP            39m
```

**Comment by david-yu** [comment]
Hi @darkn3rd Thanks for trying out Consul Service Mesh, I also responded to your query on discuss related to the naming of Consul Service Mesh. I'm interested in see how your article writeup goes. We did introduce a feature in the past where its now possible to select a single service to register in the case where multiple services select a set of pods.  See: https://www.consul.io/docs/k8s/annotations-and-labels#consul-hashicorp-com-kubernetes-service. Let me know if that works for you.

**Comment by darkn3rd** [comment]
Hi @david-yu.  For context, there are two use cases:

1. Communicate to outside via mesh: dgraph-alpha: 8080(http), 9080(grpc) and dgraph-zero: 6080(http), 5080(grpc)
2. Communicate both outside and internally, with these extra ports `dgraph-alpha`: 7080(grpc), `dgraph-zero`: 5080(grpc)

I have have done first use case on Linkerd and NGINX service mesh, but not the second one due to bugs with StatefulSet.  It looks like Linkerd fixed their issue.  I have done the second use case with Istio.  

With CSM (Consul Connect), I have added a patched the label for the headless service, and I was able to work around the issue.  But I am concerned that in general, this is a bug with Consul Connect, in that having two services (one headless REQUIRED for the StatefulSet as part of K8S docs and in practice, and one for external traffic within the K8S cluster).

I saw the link in documentation, but I have yet to fully digest the material, as this is my first time using Consul.  If I go with this approach, can all services use gRPC?  Clients will connect with either HTTP (GraphQL, DQL, other) or gRPC (DQL) to the Alpha service, and also sometimes to Zero for administrative procedures like licensing.  The Alpha and Zero nodes will also communicate between each other and are rather vociferous.
